File: examples/scripts/nithya00_constrained_opt_benchmark.cpp
/* ----------------------------------------------------------------------------
 * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
 * Atlanta, Georgia 30332-0415
 * All Rights Reserved
 * See LICENSE for the license information
 * -------------------------------------------------------------------------- */

/**
 * @file  nithya_yetong00_constrainedopt_benchmark.cpp
 * @brief Benchmark penalty method optimizer and augmented lagrangian optimizer
 * on a toy example, and output intermediate results to file.
 * @author: Nithya Jayakumar
 * @author: Yetong Zhang
 */

#include <gtsam/constrained/AugmentedLagrangianOptimizer.h>
#include <gtsam/constrained/ConstrainedOptProblem.h>
#include <gtsam/constrained/NonlinearEqualityConstraint.h>
#include <gtsam/constrained/PenaltyOptimizer.h>
#include <gtsam/nonlinear/ExpressionFactor.h>
#include <gtsam/nonlinear/expressions.h>

#include <fstream>
#include <iostream>

using namespace gtsam;

namespace {

/// Exponential function e^x.
double exp_func(const double& x, gtsam::OptionalJacobian<1, 1> H1 = {}) {
  double result = exp(x);
  if (H1) H1->setConstant(result);
  return result;
}

/// Exponential expression e^x.
Double_ exp(const Double_& x) { return Double_(exp_func, x); }

/// Pow functor used for pow function.
class PowFunctor {
 private:
  double c_;

 public:
  PowFunctor(const double& c) : c_(c) {}

  double operator()(const double& x,
                    gtsam::OptionalJacobian<1, 1> H1 = {}) const {
    if (H1) H1->setConstant(c_ * pow(x, c_ - 1));
    return pow(x, c_);
  }
};

/// Pow function.
Double_ pow(const Double_& x, const double& c) {
  PowFunctor pow_functor(c);
  return Double_(pow_functor, x);
}

/// Plus between Double expression and double.
Double_ operator+(const Double_& x, const double& d) {
  return x + Double_(d);
}

/// Negative sign operator.
Double_ operator-(const Double_& x) { return Double_(0.0) - x; }

Symbol x1_key('x', 1);
Symbol x2_key('x', 2);
Double_ x1(x1_key), x2(x2_key);

}  // namespace

int main(int argc, char** argv) {
  /// Create a constrained optimization problem with 2 cost factors and 1
  /// constraint.
  NonlinearFactorGraph graph;
  gtsam::Double_ f1 = x1 + exp(-x2);
  gtsam::Double_ f2 = pow(x1, 2.0) + 2.0 * x2 + 1.0;
  auto cost_noise = gtsam::noiseModel::Isotropic::Sigma(1, 1.0);
  graph.add(ExpressionFactor<double>(cost_noise, 0., f1));
  graph.add(ExpressionFactor<double>(cost_noise, 0., f2));

  NonlinearEqualityConstraints constraints;
  double sigma = 1.0;
  gtsam::Double_ g1 = x1 + pow(x1, 3) + x2 + pow(x2, 2);
  constraints.emplace_shared<ExpressionEqualityConstraint<double>>(
      g1, 0.0, Vector1(sigma));

  /// Create initial values.
  Values init_values;
  init_values.insert(x1_key, -0.2);
  init_values.insert(x2_key, -0.2);

  auto problem = ConstrainedOptProblem::EqConstrainedOptProblem(graph, constraints);

  /// Solve the constraint problem with Penalty Method optimizer.
  auto penalty_params = std::make_shared<gtsam::PenaltyOptimizerParams>();
  gtsam::PenaltyOptimizer penalty_optimizer(problem, init_values, penalty_params);
  Values penalty_results = penalty_optimizer.optimize();

  /// Solve the constraint problem with Augmented Lagrangian optimizer.
  auto augl_params = std::make_shared<gtsam::AugmentedLagrangianParams>();
  gtsam::AugmentedLagrangianOptimizer augl_optimizer(problem, init_values,
                                                     augl_params);
  Values augl_results = augl_optimizer.optimize();

  /// Function to evaluate constraint violation.
  auto evaluate_constraint = [&constraints](const gtsam::Values& values) {
    return constraints.violationNorm(values);
  };

  /// Function to evaluate cost.
  auto evaluate_cost = [&graph](const gtsam::Values& values) {
    double cost = graph.error(values);
    return cost;
  };

  /// Write results to files for plotting.
  std::cout << "Penalty result: cost=" << evaluate_cost(penalty_results)
            << " constraint=" << evaluate_constraint(penalty_results) << "\n";
  std::cout << "Augmented Lagrangian result: cost="
            << evaluate_cost(augl_results)
            << " constraint=" << evaluate_constraint(augl_results) << "\n";
  return 0;
}

File: examples/scripts/nithya00_constrained_opt_benchmark.py
"""
 * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
 * Atlanta, Georgia 30332-0415
 * All Rights Reserved
 * See LICENSE for the license information
 *
 * @file  nithya_yetong00_constrainedopt_benchmark.py
 * @brief Plot intermediate results for penalty method optimizer vs augmented lagrangian optimizer.
 * @author Nithya Jayakumar
 * @author Yetong Zhang
"""

import math
import numpy as np
import matplotlib.pyplot as plt 

# load optimization result data from file
def load_data(filename):
    data = []
    with open(filename) as data_file:
        lines = data_file.readlines()
        for line in lines:
            line = line.split()
            line = [float(data) for data in line]
            data.append(line)
    data = np.array(data)
    data_arranged = {}
    data_arranged["num_iters"] = data[:, 0]
    data_arranged["cumulative_iters"] = np.cumsum(data[:, 0])
    data_arranged["mu"] = data[:, 1]
    data_arranged["feasibility"] = data[:, 2]
    data_arranged["optimality"] = data[:, 3]
    return data_arranged

# create 2 plots: 
# plot1: penalty parameter vs. cumulative lm iterations
# plot2: feasiblity, optimality vs. cumulative lm iterations
def make_plot(data):
    # fig = plt.figure(figsize=(10, 10), dpi=160)
    fig, axes = plt.subplots(1, 2, figsize=(12, 5))

    axes[0].set_yscale('log')
    axes[0].plot(data["cumulative_iters"], data["mu"])
    axes[0].set_xlabel("Cumulative L-M Iterations")
    axes[0].set_ylabel("Penalty Parameter mu")

    axes[1].set_xlabel("Cumulative L-M Iterations")
    axes[1].set_ylabel("Residual")
    axes[1].plot(data["cumulative_iters"], data["feasibility"], label = "Feasibility", color="b")
    axes[1].plot(data["cumulative_iters"], data["optimality"], label = "Optimality", color="r")
    axes[1].legend()

# Load data for penalty method optimizer.
data_penalty = load_data('build/scripts/penalty_data.txt')
data_augl = load_data('build/scripts/augl_data.txt')

make_plot(data_penalty)
make_plot(data_augl)
plt.show()
